packages/expo/cli/start/server/BundlerDevServer.ts

```diff
@@ -0,0 +1,308 @@
+
+  /** Should instruct the bundler to create minified bundles. */
+  minify?: boolean;
+
+  // Webpack options
+  /** Should modify and create PWA icons. */
+  isImageEditingEnabled?: boolean;
+
+  location: CreateURLOptions;
+}
+
+const PLATFORM_MANAGERS = {
+  simulator: ApplePlatformManager,
+  emulator: AndroidPlatformManager,
+};
+
+const MIDDLEWARES = {
+  classic: ClassicManifestMiddleware,
+  'expo-updates': ExpoGoManifestHandlerMiddleware,
+};
+
+export abstract class BundlerDevServer {
+  /** Name of the bundler. */
+  abstract get name(): string;
+
+  /** Ngrok instance for managing tunnel connections. */
+  protected ngrok: AsyncNgrok | null = null;
+  /** Interfaces with the Expo 'Development Session' API. */
+  protected devSession: DevelopmentSession | null = null;
+  /** Http server and related info. */
+  protected instance: DevServerInstance | null = null;
+  /** Native platform interfaces for opening projects.  */
+  private platformManagers: Record<string, PlatformManager<any>> = {};
+  /** Manages the creation of dev server URLs. */
+  protected urlCreator?: UrlCreator | null = null;
+
+  constructor(
+    /** Project root folder. */
+    public projectRoot: string,
+    // TODO: Replace with custom scheme maybe...
+    public isDevClient?: boolean
+  ) {}
+
+  protected setInstance(instance: DevServerInstance) {
+    this.instance = instance;
+  }
+
+  /** Get the manifest middleware function. Exposed for testing. */
+  public _getManifestMiddleware(
+    options: Pick<BundlerStartOptions, 'minify' | 'mode' | 'forceManifestType'> = {}
+  ) {
+    const manifestType = options.forceManifestType || 'classic';
+    const Middleware = MIDDLEWARES[manifestType];
+    assert(Middleware, `Manifest middleware for type '${manifestType}' not found`);
+
+    const urlCreator = this.getUrlCreator();
+    const middleware = new Middleware(this.projectRoot, {
+      constructUrl: urlCreator.constructUrl.bind(urlCreator),
+      mode: options.mode,
+      minify: options.minify,
+      isNativeWebpack: this.name === 'webpack' && this.isTargetingNative(),
+    });
+    return middleware.getHandler();
+  }
+
+  /** Start the dev server using settings defined in the start command. */
+  public abstract startAsync(options: BundlerStartOptions): Promise<DevServerInstance>;
+
+  protected async postStartAsync(options: BundlerStartOptions) {
+    if (options.location.hostType === 'tunnel' && !APISettings.isOffline) {
+      await this._startTunnelAsync();
+    }
+    await this.startDevSessionAsync();
+
+    this.watchConfig();
+  }
+
+  protected abstract getConfigModuleIds(): string[];
+
+  protected watchConfig() {
+    const notifier = new FileNotifier(this.projectRoot, this.getConfigModuleIds());
+    notifier.startObserving();
+  }
+
+  /** Create ngrok instance and start the tunnel server. Exposed for testing. */
+  public async _startTunnelAsync(): Promise<AsyncNgrok | null> {
+    const port = this.getInstance()?.location?.port;
```

Comment:
location is non-null:
```suggestion
    const port = this.getInstance()?.location.port;
```